jujugui/static/gui/src/app/components/deployment-flow/sshkey/sshkey.js

```diff
@@ -30,22 +15,298 @@ with this program.  If not, see <http://www.gnu.org/licenses/>.
+  _addGithubKeysCallback(error, keys) {
+    if (error) {
+      console.error(error);
+      this.setState({error: error});
+      return;
+    }
+
+    if (keys.length === 0) {
+      this.setState({error:
+        (<span>No keys found. <a className="link"
+          href="https://github.com/settings/keys" target="_blank">
+          Create an SSH Key</a>.</span>)});
+      return;
+    }
+
+    let SSHkeys = this.state.SSHkeys;
+    keys.forEach(key => {
+      if (!this._keyExists(key)) {
+        SSHkeys.push(key);
+      }
+    });
+    this.props.setSSHKey(SSHkeys[0].text);
+    this.setState({SSHkeys: SSHkeys, buttonDisabled: true});
+    this.refs.githubUsername.setValue(null);
+    this.refs.githubUsername.focus();
+  }
+
+  /**
+    Handle clicking AddMoreKeys.
+  */
+  _handleAddMoreKeys() {
+    if (this.state.buttonDisabled) {
+      return;
+    }
+    const source = this.refs.sshSource.getValue();
+    this.setState({error: null});
+    if (source === 'github') {
+      const githubUsername = this.refs.githubUsername.getValue();
+      this.props.githubSSHKeys(
+        new yui.juju.environments.web.WebHandler,
+        githubUsername,
+        this._addGithubKeysCallback.bind(this)
+      );
+    } else if (source === 'manual') {
+      const manualKey = this.refs.sshKey.getValue();
+      const key = this._splitKey(manualKey);
+      if (key) {
+        this.props.setSSHKey(manualKey);
+        let SSHkeys = this.state.SSHkeys;
+        if (!this._keyExists(key)) {
+          SSHkeys.push(key);
+        }
+        this.setState({SSHkeys: SSHkeys, buttonDisabled: true});
+        this.refs.sshKey.setValue(null);
+        this.refs.sshKey.focus();
+      }
+    }
+  }
+
+  /**
+    Handle source change.
+  */
+  _handleSourceChange() {
+    const source = this.refs.sshSource.getValue();
+    this.setState({addSource: source, buttonDisabled: true});
+  }
+
+  /**
+    Remove key from table.
+
+    @param {Number} keyId The key's ID.
+  */
+  _removeKey(keyId) {
+    const newSSHkeyList = this.state.SSHkeys.filter(key => {
+      return key.id !== keyId;
+    });
+
+    this.setState({
+      SSHkeys: newSSHkeyList
+    });
+
+    if (!newSSHkeyList.length) {
+      this.props.setSSHKey(null);
+    }
+  }
+
+  /**
+    Create the added keys section.
+
+    @return {Object} The React list of keys.
+  */
+  _generateAddedKeys() {
+    const SSHkeys = this.state.SSHkeys;
+    const stringLengths = 30;
+
+    if (Object.keys(SSHkeys).length === 0) {
+      return false;
+    }
+
+    let listBody = [];
+    SSHkeys.forEach((key, i) => {
+      let uniqueKey = key.id + i;
+      let body = key.body;
+
+      if (body.length >= stringLengths * 2) {
+        const bodyStart = key.body.substring(0, stringLengths);
+        const bodyEnd = key.body.substring(key.body.length - stringLengths);
+        body = `${bodyStart}...${bodyEnd}`;
+      }
+      listBody.push(
+        <li className="deployment-flow__row twelve-col" key={uniqueKey}>
```

Comment:
lets use css grid instead of these `col` classes for the whole component - it really cleans up the markup and it's a standard